### PR TITLE
feat(ledger): add tax code and tax behavior to ledger types

### PR DESCRIPTION
**Pull request by mark-vass-konghq**
## Summary by CodeRabbit

* **New Features**
  * Added optional tax behavior (inclusive/exclusive) on routes, persisted and incorporated into routing keys; tax code and tax behavior now propagate through routing and transaction flows.

* **Refactor**
  * Routing/filter types now preserve presence semantics to distinguish absent vs explicitly null tax fields.

* **Bug Fixes**
  * Sub-account resolution and routing correctly differentiate nil vs non-nil tax codes and tax behavior.

* **Tests**
  * Added coverage for tax behavior validation, routing-key variants, normalization, pairing keys, and tax-code isolation.

* **Chores**
  * DB migration to add nullable tax_behavior and backfill routing keys; down-migration removes added segment.

[Review Change Stack](https://app.coderabbit.ai/change-stack/openmeterio/openmeter/pull/4291)

### CODE REVIEW

File: openmeter/ledger/routing.go [hunk 1]
```diff
@@ -270,6 +313,7 @@ func BuildRoutingKeyV1(route Route) (RoutingKey, error) {
 	value := strings.Join([]string{
 		"currency:" + string(normalizedRoute.Currency),
 		"tax_code:" + optionalStringValue(normalizedRoute.TaxCode),
+		"tax_behavior:" + optionalTaxBehaviorValue(normalizedRoute.TaxBehavior),
```
Comment (GAlexIHU) [inline openmeter/ledger/routing.go]: pls advise, if we just add it like this on the same version we break all envs where its deployed

File: openmeter/ledger/primitives.go [hunk 2]
```diff
@@ -49,11 +49,12 @@ type SubAccount interface {
 type RouteFilter struct {
 	Currency currencyx.Code
 
-	// DEFERRED: tax/feature not active yet.
 	// Non-currency fields are retained for near-future expansion.
-	TaxCode   *string
-	Features  []string
-	CostBasis mo.Option[*alpacadecimal.Decimal]
+	// mo.Option provides nil-distinction: absent = don't filter, Some(nil) = filter for null, Some(&v) = filter for v.
```
Comment (GAlexIHU) [inline openmeter/ledger/primitives.go]: no need for that comment imo

**Review by GAlexIHU — CHANGES_REQUESTED** [review]
thx, if you don't mind I'd prefer if a bit more stuff was collected on the branch that actually do something (ik unpopular opinion)

**Comment by mark-vass-konghq** [comment]
> thx, if you don't mind I'd prefer if a bit more stuff was collected on the branch that actually do something (ik unpopular opinion)

What changes needed to be done to not break our env?

File: openmeter/ent/schema/ledger_account.go [hunk 3]
```diff
@@ -116,6 +116,9 @@ func (LedgerSubAccountRoute) Fields() []ent.Field {
 		// Literal routing values
 		field.String("currency").Immutable(),
 		field.String("tax_code").Optional().Nillable().Immutable(),
```
Comment (GAlexIHU) [inline openmeter/ent/schema/ledger_account.go:119]: is this an ID? is this an FK? what is this? what is this meant to be?

File: openmeter/ledger/transactions/testenv_test.go [hunk 4]
```diff
@@ -101,3 +101,43 @@ func (e *transactionsTestEnv) fundPriorityWithCostBasis(t *testing.T, priority i
...
 }
+
+func (e *transactionsTestEnv) fundPriorityWithCostBasisAndTaxCode(t *testing.T, priority int, amount int64, costBasis *alpacadecimal.Decimal, taxCode *string) ledger.SubAccount {
+	t.Helper()
+
+	subAccount, err := e.CustomerAccounts.FBOAccount.GetSubAccountForRoute(t.Context(), ledger.CustomerFBORouteParams{
+		Currency:       e.Currency,
+		CostBasis:      costBasis,
+		CreditPriority: priority,
+		TaxCode:        taxCode,
```
Comment (GAlexIHU) [inline openmeter/ledger/transactions/testenv_test.go]: see you've added TaxCode, if you've added TaxBehavior in the schema maybe add that too at this poing

**Review by GAlexIHU — CHANGES_REQUESTED** [review]

File: openmeter/ledger/routing.go (hunk 1 above)
Comment (GAlexIHU) [inline openmeter/ledger/routing.go]: as this is deployed this would break all routing keys. thought excercise: would you migrate existing keys or move to v2? (i mean thats the reason we have versioning)

File: tools/migrate/migrations/20260506103900_add_ledger_tax_behavior.up.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
+-- modify "ledger_sub_account_routes" table
+ALTER TABLE "ledger_sub_account_routes" ADD COLUMN "tax_behavior" character varying NULL;
+-- Backfill routing_key for existing V1 rows: insert tax_behavior:null segment between tax_code and features.
+-- Pre-existing rows were written without tax_behavior in the key; without this they would not be found by new lookups.
+UPDATE "ledger_sub_account_routes"
+SET "routing_key" = regexp_replace(
+    "routing_key",
+    '(\|tax_code:[^|]*)(\|features:)',
+    '\1|tax_behavior:null\2'
+)
```
Comment (GAlexIHU) [inline tools/migrate/migrations/20260506103900_add_ledger_tax_behavior.up.sql]: ok i see you chose this: you should add a migration test with "realistic" seed  data

(also as this is a data migration, if you haven't yet, please spend some minutes thinking through how this can go wrong)

**Review by GAlexIHU — CHANGES_REQUESTED** [review]
I need you to write e2e assertions that verify tax config is from charge to earnings

**Comment by GAlexIHU** [comment]
also routingrules need to be updated

File: test/credits/sanity_test.go [hunk 6]
```diff
@@ -1422,3 +1424,130 @@ func (s *SanitySuite) TestFlatFeeCreditOnlySanity() {
 		assertDelta("external wash after later purchase settlement", start.externalWash, alpacadecimal.NewFromInt(-50), s.MustWashBalance(ns, USD, mo.Some(&externalCostBasis)))
 	})
 }
+
+// TestTaxCodeFlowsFromCreditPurchaseToEarnings verifies the end-to-end routing of
+// TaxCode: credit purchase → FBO sub-account → accrued → earnings.
+// Credits funded with a TaxCode must land in a TaxCode-keyed earnings sub-account
+// after charge collection and revenue recognition.
+func (s *SanitySuite) TestTaxCodeFlowsFromCreditPurchaseToEarnings() {
```
Comment (GAlexIHU) [inline test/credits/sanity_test.go:3205]: its good you tested for credits. what do you think about invoices?

File: openmeter/ledger/chargeadapter/usagebased.go [hunk 7]
```diff
@@ -131,6 +131,7 @@ func (h *usageBasedHandler) OnPaymentAuthorized(ctx context.Context, input usage
 			At:        eventTime,
 			Amount:    receivableReplenishment,
 			Currency:  input.Charge.Intent.Currency,
+			TaxCode:   taxCodeIDFromIntent(input.Charge.Intent.TaxConfig),
```
Comment (GAlexIHU) [inline openmeter/ledger/chargeadapter/usagebased.go]: i think this splits routes. invoice accrual still writes receivable/accrued without tax_code, but auth/settlement now use taxCodeIDFromIntent. for tax-configured usage charges that means nil-tax-code accrual + tax-code payment path, so the original receivable may never clear and earnings can land in the wrong bucket. can we thread the same tax code through accrual too?

File: test/credits/sanity_test.go [hunk 8]
```diff
@@ -1422,3 +1424,130 @@ func (s *SanitySuite) TestFlatFeeCreditOnlySanity() {
...
+		s.Equal(float64(0), s.MustCustomerFBOBalanceForTaxCode(cust.GetID(), USD, mo.Some(&nilCostBasis), mo.Some[*string](nil)).InexactFloat64(),
+			"nil-TaxCode FBO sub-account must be untouched")
+	})
+
+	var flatFeeChargeID string
+	s.Run("create and advance flat fee credit-only charge", func() {
+		res, err := s.Charges.Create(ctx, charges.CreateInput{
+			Namespace: ns,
+			Intents: charges.ChargeIntents{
+				s.CreateMockChargeIntent(CreateMockChargeIntentInput{
```
Comment (GAlexIHU) [inline test/credits/sanity_test.go:3262]: this test is close but i dont think it proves the thing i asked for. the tax config is on the promotional credit funding, not the charge being consumed. so were proving funded credit tax code survives to earnings, not charge tax config -> earnings. can we put tax config on the flat fee/usage charge and assert that path?

File: openmeter/ledger/transactions/earnings.go [hunk 9]
```diff
@@ -136,6 +137,7 @@ func (t RecognizeEarningsFromAttributableAccruedTemplate) resolveEarningsSubAccB
 			// same route. Recognition only needs one earnings destination per route.
 			earnings, err := earningsAccount.GetSubAccountForRoute(ctx, ledger.BusinessRouteParams{
 				Currency:  t.Currency,
+				TaxCode:   collection.subAccount.Route().TaxCode,
```
Comment (GAlexIHU) [inline openmeter/ledger/transactions/earnings.go]: this relies on tax_code being preserved, but the routing rules dont enforce it. fbo -> accrued and accrued -> earnings still only match currency/cost_basis. can we include RouteFieldTaxCode there + tests? otherwise future changes can silently dump tax-coded credits into nil-tax-code earnings.

**Review by GAlexIHU — CHANGES_REQUESTED** [review]
ai was (not) (ab)used when writing these comments

File: openmeter/ledger/transactions/accrual.go [hunk 10]
```diff
@@ -213,12 +230,20 @@ func costBasisKey(costBasis *alpacadecimal.Decimal) string {
...
+	return *taxCode
+}
+
 // TransferCustomerFBOAdvanceToAccruedTemplate moves value from the synthetic advance-backed
 // customer FBO route into the matching accrued route.
 type TransferCustomerFBOAdvanceToAccruedTemplate struct {
 	At             time.Time
 	Amount         alpacadecimal.Decimal
 	Currency       currencyx.Code
+	TaxCode        *string
```
Comment (GAlexIHU) [inline openmeter/ledger/transactions/accrual.go:247]: wouldn't you also need TaxBehavior?

File: openmeter/ledger/transactions/accrual.go [hunk 11]
```diff
@@ -146,6 +147,21 @@ func (t TransferCustomerFBOToAccruedTemplate) resolve(ctx context.Context, custo
...
 }
 
+// resolveAccruedSubAccByRoutePairingKey routes each FBO source to its
+// matching accrued sub-account using the SOURCE route's TaxCode and
+// CostBasis. Charge intent TaxConfig does NOT override the destination
+// bucket on this path: accrual preserves funding-source attribution by
+// design (credit-only flows do not translate TaxCode at consumption).
+//
+// If charge.Intent.TaxConfig diverges from FBO source TaxCode, the
+// resulting accrued posting lands in the FUNDING source bucket. The
```
Comment (GAlexIHU) [inline openmeter/ledger/transactions/accrual.go]: are we sure this whole long comment belongs here?

**Review by GAlexIHU — CHANGES_REQUESTED** [review]
I think you forgot about TaxBehavior. I didn't go to find the exact lines in sanity where you miss asserting for it, but its missed. Please add & fix

File: openmeter/ledger/chargeadapter/creditpurchase.go [hunk 12]
```diff
@@ -201,57 +203,52 @@ func (h *creditPurchaseHandler) issueCreditPurchaseGroup(ctx context.Context, ch
...
-			Currency:      charge.Intent.Currency,
-			FromCostBasis: nil,
-			ToCostBasis:   &costBasis,
-		})
+		if attribution.accruedAmount.IsPositive() {
+			templates = append(templates, transactions.TranslateCustomerAccruedCostBasisTemplate{
+				At:            charge.CreatedAt,
+				Amount:        attribution.accruedAmount,
+				Currency:      charge.Intent.Currency,
+				TaxCode:       attribution.taxCode,
```
Comment (GAlexIHU) [inline openmeter/ledger/chargeadapter/creditpurchase.go:238]: Business invariant here is source usage tax config, not credit-purchase tax config. Credit-only usage creates the advance accrued bucket with both TaxCode and TaxBehavior, but this backfill path only carries TaxCode: advanceAttribution has no taxBehavior, unattributedAccruedByTaxCode groups by tax code only, and this TranslateCustomerAccruedCostBasisTemplate resolves a nil TaxBehavior route. If two advance buckets share a tax code but differ by inclusive/exclusive behavior, the backfill can merge them and move value through the nil-behavior accrued route instead of preserving the usage charge tax behavior. Please carry TaxBehavior in the attribution key/value and pass it into this template.

File: openmeter/ledger/accounts.go [hunk 13]
```diff
@@ -128,8 +131,10 @@ type BusinessAccount interface {
 }
 
 type BusinessRouteParams struct {
-	Currency  currencyx.Code
-	CostBasis *alpacadecimal.Decimal
+	Currency    currencyx.Code
+	TaxCode     *string
```
Comment (GAlexIHU) [inline openmeter/ledger/accounts.go:135]: For earnings, you should make TC & TB required in the RouteParams + also have routevalidations for it

File: openmeter/ledger/transactions/customer.go [hunk 14]
```diff
@@ -125,6 +126,7 @@ func (t IssueCustomerReceivableTemplate) resolve(ctx context.Context, customerID
 
 	rec, err := customerAccounts.ReceivableAccount.GetSubAccountForRoute(ctx, ledger.CustomerReceivableRouteParams{
 		Currency:                       t.Currency,
+		TaxCode:                        t.TaxCode,
```
Comment (GAlexIHU) [inline openmeter/ledger/transactions/customer.go]: Business issue: credit-purchase `TaxConfig` is purchase-side tax metadata, not the revenue-recognition tax dimension.

Right now this PR still threads credit-purchase tax config into ledger routes. That makes the ledger able to say “this earning belongs to the purchase tax code”, which is not the thing we care about. The tax code we need on accrued/earnings is the tax config of the FF/UBP charge that actually realized the usage/earning.

Please remove credit-purchase `TaxCode` / `TaxBehavior` propagation into ledger route dimensions. Purchased credit issuance should stay tax-neutral from the revenue-recognition POV; tax dimensions should enter when usage accrues.

File: openmeter/ledger/routingrules/routingrules.go [hunk 15]
```diff
@@ -197,6 +198,33 @@ func (r RequireSameRouteRule) Validate(tx TxView) error {
 	return requireMatchingRouteFields(tx.EntriesOf(r.Left), tx.EntriesOf(r.Right), r.Left, r.Right, r.Fields)
 }
 
+type RequireTaxDimensionScopeRule struct{}
+
+func (r RequireTaxDimensionScopeRule) Validate(tx TxView) error {
+	for _, entry := range tx.Entries() {
+		if entry.AccountType() == ledger.AccountTypeCustomerFBO && entry.Route().TaxCode != nil {
```
Comment (GAlexIHU) [inline openmeter/ledger/routingrules/routingrules.go]: I'd rather you handle this all in the type switch below. also, if you forbid it on one then please forbid it on all where it isnt allowed.

**Review by GAlexIHU — CHANGES_REQUESTED** [review]

**Review by GAlexIHU — APPROVED** [review]
